server/src/main/java/thesistrack/ls1/controller/ThesisController.java

```diff
@@ -347,23 +352,97 @@ public ResponseEntity<ThesisDto> deletePresentation(@PathVariable UUID thesisId,
     }
 
     @PostMapping("/{thesisId}/comments")
-    public ResponseEntity<ThesisCommentDto> createComment(@PathVariable UUID thesisId) {
-        throw new ResponseStatusException(HttpStatus.NOT_IMPLEMENTED, "This feature is not implemented yet");
+    public ResponseEntity<ThesisCommentDto> createComment(
+            @PathVariable UUID thesisId,
+            @RequestPart("data") PostThesisCommentPayload payload,
+            @RequestPart(value = "file", required = false) MultipartFile file,
+            JwtAuthenticationToken jwt
+    ) {
+        User authenticatedUser = authenticationService.getAuthenticatedUser(jwt);
+        Thesis thesis = thesisService.findById(thesisId);
+
+        if (payload.commentType() == ThesisCommentType.ADVISOR && !thesis.hasAdvisorAccess(authenticatedUser)) {
+            throw new AccessDeniedException("You need to be a advisor of this thesis to perform this action");
+        }
+
+        if (!thesis.hasStudentAccess(authenticatedUser)) {
+            throw new AccessDeniedException("You need to be a student of this thesis to perform this action");
+        }
+
+        ThesisComment comment = thesisCommentService.postComment(
+                authenticatedUser,
+                thesis,
+                RequestValidator.validateNotNull(payload.commentType()),
+                RequestValidator.validateStringMaxLength(payload.message(), StringLimits.LONGTEXT.getLimit()),
+                file
+        );
+
+        return ResponseEntity.ok(ThesisCommentDto.fromCommentEntity(comment));
     }
 
-    @GetMapping("/{thesisId}/comments/{commentId}")
-    public ResponseEntity<Resource> getCommentFile(@PathVariable UUID thesisId, @PathVariable UUID commentId) {
-        throw new ResponseStatusException(HttpStatus.NOT_IMPLEMENTED, "This feature is not implemented yet");
+    @GetMapping("/{thesisId}/comments/{commentId}/file")
+    public ResponseEntity<Resource> getCommentFile(
+            @PathVariable UUID thesisId,
+            @PathVariable UUID commentId,
+            JwtAuthenticationToken jwt
+    ) {
+        User authenticatedUser = authenticationService.getAuthenticatedUser(jwt);
+        ThesisComment comment = thesisCommentService.findById(thesisId, commentId);
+
+        if (comment.getType() == ThesisCommentType.ADVISOR && !comment.getThesis().hasAdvisorAccess(authenticatedUser)) {
+            throw new AccessDeniedException("You need to be a advisor of this thesis to perform this action");
+        }
+
+        if (!comment.getThesis().hasReadAccess(authenticatedUser)) {
+            throw new AccessDeniedException("You do not have the required permissions to view this thesis");
+        }
+
+        return ResponseEntity.ok()
+                .contentType(MediaType.APPLICATION_PDF)
+                .header(HttpHeaders.CONTENT_DISPOSITION, String.format("attachment; filename=comment_%s.pdf", commentId))
```

Comment:
Could we use the original filename here? That would allow both advisor and student to know what they are about to download and what they already posted 👍

I'd still recommend a sanitization step that, e.g., only allows `pdf`, `docx`, `tex`, etc

Reply:
I will create a ticket for that in the backlog. Right now the original filename is not preserved when uploading files to thesis track. Adding that would require larger changes I would not want to do in this PR